// @backstage/plugin-catalog/dist/components/HasSubcomponentsCard/HasSubcomponentsCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { componentEntityColumns, asComponentEntities } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasSubcomponentsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const {
    variant = "gridItem",
    tableOptions = {},
    title = t("hasSubcomponentsCard.title")
  } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Component",
      relationType: RELATION_HAS_PART,
      columns: componentEntityColumns,
      asRenderableEntities: asComponentEntities,
      emptyMessage: t("hasSubcomponentsCard.emptyMessage"),
      emptyHelpLink: "https://backstage.io/docs/features/software-catalog/descriptor-format#specsubcomponentof-optional",
      tableOptions
    }
  );
}

export { HasSubcomponentsCard };
