// @backstage/core-compat-api/dist/convertLegacyRouteRef.esm.js
import { routeRefType } from './core-plugin-api/src/routing/types.esm.js';
import { createRouteRef, createSubRouteRef, createExternalRouteRef } from '@backstage/frontend-plugin-api';
import { toInternalRouteRef } from './frontend-plugin-api/src/routing/RouteRef.esm.js';
import { toInternalSubRouteRef } from './frontend-plugin-api/src/routing/SubRouteRef.esm.js';
import { toInternalExternalRouteRef } from './frontend-plugin-api/src/routing/ExternalRouteRef.esm.js';

function convertLegacyRouteRefs(refs) {
  return Object.fromEntries(
    Object.entries(refs).map(([name, ref]) => [
      name,
      convertLegacyRouteRef(ref)
    ])
  );
}
function convertLegacyRouteRef(ref) {
  if ("$$type" in ref) {
    return ref;
  }
  const type = ref[routeRefType];
  if (type === "absolute") {
    const legacyRef = ref;
    const legacyRefStr = String(legacyRef);
    const newRef = toInternalRouteRef(
      createRouteRef({
        params: legacyRef.params
      })
    );
    return Object.assign(legacyRef, {
      $$type: "@backstage/RouteRef",
      version: "v1",
      T: newRef.T,
      getParams() {
        return newRef.getParams();
      },
      getDescription() {
        return legacyRefStr;
      },
      setId(id) {
        newRef.setId(id);
      },
      toString() {
        return legacyRefStr;
      }
    });
  }
  if (type === "sub") {
    const legacyRef = ref;
    const legacyRefStr = String(legacyRef);
    const newRef = toInternalSubRouteRef(
      createSubRouteRef({
        path: legacyRef.path,
        parent: convertLegacyRouteRef(legacyRef.parent)
      })
    );
    return Object.assign(legacyRef, {
      $$type: "@backstage/SubRouteRef",
      version: "v1",
      T: newRef.T,
      getParams() {
        return newRef.getParams();
      },
      getParent() {
        return newRef.getParent();
      },
      getDescription() {
        return legacyRefStr;
      },
      toString() {
        return legacyRefStr;
      }
    });
  }
  if (type === "external") {
    const legacyRef = ref;
    const legacyRefStr = String(legacyRef);
    const newRef = toInternalExternalRouteRef(
      createExternalRouteRef({
        params: legacyRef.params,
        optional: legacyRef.optional,
        defaultTarget: "getDefaultTarget" in legacyRef ? legacyRef.getDefaultTarget() : void 0
      })
    );
    return Object.assign(legacyRef, {
      $$type: "@backstage/ExternalRouteRef",
      version: "v1",
      T: newRef.T,
      optional: newRef.optional,
      getParams() {
        return newRef.getParams();
      },
      getDescription() {
        return legacyRefStr;
      },
      // This might already be implemented in the legacy ref, but we override it just to be sure
      getDefaultTarget() {
        return newRef.getDefaultTarget();
      },
      setId(id) {
        newRef.setId(id);
      },
      toString() {
        return legacyRefStr;
      }
    });
  }
  throw new Error(`Failed to convert legacy route ref, unknown type '${type}'`);
}

export { convertLegacyRouteRef, convertLegacyRouteRefs };
