// @backstage/plugin-auth-backend-module-onelogin-provider/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var passportOneloginOauth = require('passport-onelogin-oauth');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var backendPluginApi = require('@backstage/backend-plugin-api');

const oneLoginAuthenticator = pluginAuthNode.createOAuthAuthenticator({
  defaultProfileTransform: pluginAuthNode.PassportOAuthAuthenticatorHelper.defaultProfileTransform,
  initialize({ callbackUrl, config }) {
    const clientId = config.getString("clientId");
    const clientSecret = config.getString("clientSecret");
    const issuer = config.getString("issuer");
    return pluginAuthNode.PassportOAuthAuthenticatorHelper.from(
      new passportOneloginOauth.Strategy(
        {
          clientID: clientId,
          clientSecret,
          callbackURL: callbackUrl,
          issuer,
          passReqToCallback: false
        },
        (accessToken, refreshToken, params, fullProfile, done) => {
          done(
            void 0,
            { fullProfile, params, accessToken },
            { refreshToken }
          );
        }
      )
    );
  },
  async start(input, helper) {
    input.scope = "openid";
    return helper.start(input, {
      accessType: "offline",
      prompt: "consent"
    });
  },
  async authenticate(input, helper) {
    return helper.authenticate(input);
  },
  async refresh(input, helper) {
    input.scope = "openid";
    return helper.refresh(input);
  }
});

exports.oneLoginSignInResolvers = void 0;
((oneLoginSignInResolvers2) => {
  oneLoginSignInResolvers2.usernameMatchingUserEntityName = pluginAuthNode.createSignInResolverFactory({
    create() {
      return async (info, ctx) => {
        const { result } = info;
        const id = result.fullProfile.username;
        if (!id) {
          throw new Error(`OneLogin user profile does not contain a username`);
        }
        return ctx.signInWithCatalogUser({ entityRef: { name: id } });
      };
    }
  });
})(exports.oneLoginSignInResolvers || (exports.oneLoginSignInResolvers = {}));

const authModuleOneLoginProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "onelogin-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint
      },
      async init({ providers }) {
        providers.registerProvider({
          providerId: "onelogin",
          factory: pluginAuthNode.createOAuthProviderFactory({
            authenticator: oneLoginAuthenticator,
            signInResolverFactories: {
              ...exports.oneLoginSignInResolvers,
              ...pluginAuthNode.commonSignInResolvers
            }
          })
        });
      }
    });
  }
});

exports.default = authModuleOneLoginProvider;
exports.oneLoginAuthenticator = oneLoginAuthenticator;
