// @backstage/core-components/dist/components/TabbedLayout/RoutedTabs.esm.js
import React__default, { useMemo } from 'react';
import { Helmet } from 'react-helmet';
import { useNavigate, useParams, useRoutes, matchRoutes } from 'react-router-dom';
import { Content } from '../../layout/Content/Content.esm.js';
import { HeaderTabs } from '../../layout/HeaderTabs/HeaderTabs.esm.js';

function useSelectedSubRoute(subRoutes) {
  const params = useParams();
  const routes = subRoutes.map(({ path, children }) => ({
    caseSensitive: false,
    path: `${path}/*`,
    element: children
  }));
  const sortedRoutes = routes.sort(
    (a, b) => (
      // remove "/*" symbols from path end before comparing
      b.path.replace(/\/\*$/, "").localeCompare(a.path.replace(/\/\*$/, ""))
    )
  );
  const element = useRoutes(sortedRoutes) ?? subRoutes[0]?.children;
  let currentRoute = params["*"] ?? "";
  if (!currentRoute.startsWith("/")) {
    currentRoute = `/${currentRoute}`;
  }
  const [matchedRoute] = matchRoutes(sortedRoutes, currentRoute) ?? [];
  const foundIndex = matchedRoute ? subRoutes.findIndex((t) => `${t.path}/*` === matchedRoute.route.path) : 0;
  return {
    index: foundIndex === -1 ? 0 : foundIndex,
    element,
    route: subRoutes[foundIndex] ?? subRoutes[0]
  };
}
function RoutedTabs(props) {
  const { routes } = props;
  const navigate = useNavigate();
  const { index, route, element } = useSelectedSubRoute(routes);
  const headerTabs = useMemo(
    () => routes.map((t) => ({
      id: t.path,
      label: t.title,
      tabProps: t.tabProps
    })),
    [routes]
  );
  const onTabChange = (tabIndex) => {
    let { path } = routes[tabIndex];
    path = path.replace(/\/\*$/, "");
    path = path.replace(/^\//, "");
    navigate(path);
  };
  return /* @__PURE__ */ React__default.createElement(React__default.Fragment, null, /* @__PURE__ */ React__default.createElement(
    HeaderTabs,
    {
      tabs: headerTabs,
      selectedIndex: index,
      onChange: onTabChange
    }
  ), /* @__PURE__ */ React__default.createElement(Content, null, /* @__PURE__ */ React__default.createElement(Helmet, { title: route?.title }), element));
}

export { RoutedTabs, useSelectedSubRoute };
