// @backstage/plugin-search-react/dist/alpha.esm.js
import React, { lazy } from 'react';
import { createExtensionDataRef, createSchemaFromZod, createExtension, ExtensionBoundary } from '@backstage/frontend-plugin-api';
import { SearchResultListItemExtension } from './extensions.esm.js';

function createSearchResultListItemExtension(options) {
  const configSchema = "configSchema" in options ? options.configSchema : createSchemaFromZod(
    (z) => z.object({
      noTrack: z.boolean().default(false)
    })
  );
  return createExtension({
    kind: "search-result-list-item",
    namespace: options.namespace,
    name: options.name,
    attachTo: options.attachTo ?? {
      id: "page:search",
      input: "items"
    },
    configSchema,
    output: {
      item: createSearchResultListItemExtension.itemDataRef
    },
    factory({ config, node }) {
      const ExtensionComponent = lazy(
        () => options.component({ config }).then((component) => ({ default: component }))
      );
      return {
        item: {
          predicate: options.predicate,
          component: (props) => /* @__PURE__ */ React.createElement(ExtensionBoundary, { node }, /* @__PURE__ */ React.createElement(
            SearchResultListItemExtension,
            {
              rank: props.rank,
              result: props.result,
              noTrack: config.noTrack
            },
            /* @__PURE__ */ React.createElement(ExtensionComponent, { ...props })
          ))
        }
      };
    }
  });
}
((createSearchResultListItemExtension2) => {
  createSearchResultListItemExtension2.itemDataRef = createExtensionDataRef("search.search-result-list-item.item");
})(createSearchResultListItemExtension || (createSearchResultListItemExtension = {}));

export { createSearchResultListItemExtension };
