// @backstage/plugin-catalog/dist/alpha/entityCards.esm.js
import React from 'react';
import { createEntityCardExtension } from '@backstage/plugin-catalog-react/alpha';
import { compatWrapper } from '@backstage/core-compat-api';

const catalogAboutEntityCard = createEntityCardExtension({
  name: "about",
  loader: async () => import('../components/AboutCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.AboutCard, { variant: "gridItem" }))
  )
});
const catalogLinksEntityCard = createEntityCardExtension({
  name: "links",
  filter: "has:links",
  loader: async () => import('../components/EntityLinksCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.EntityLinksCard, { variant: "gridItem" }))
  )
});
const catalogLabelsEntityCard = createEntityCardExtension({
  name: "labels",
  filter: "has:labels",
  loader: async () => import('../components/EntityLabelsCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.EntityLabelsCard, { variant: "gridItem" }))
  )
});
const catalogDependsOnComponentsEntityCard = createEntityCardExtension({
  name: "depends-on-components",
  loader: async () => import('../components/DependsOnComponentsCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.DependsOnComponentsCard, { variant: "gridItem" }))
  )
});
const catalogDependsOnResourcesEntityCard = createEntityCardExtension({
  name: "depends-on-resources",
  loader: async () => import('../components/DependsOnResourcesCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.DependsOnResourcesCard, { variant: "gridItem" }))
  )
});
const catalogHasComponentsEntityCard = createEntityCardExtension({
  name: "has-components",
  loader: async () => import('../components/HasComponentsCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.HasComponentsCard, { variant: "gridItem" }))
  )
});
const catalogHasResourcesEntityCard = createEntityCardExtension({
  name: "has-resources",
  loader: async () => import('../components/HasResourcesCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.HasResourcesCard, { variant: "gridItem" }))
  )
});
const catalogHasSubcomponentsEntityCard = createEntityCardExtension({
  name: "has-subcomponents",
  loader: async () => import('../components/HasSubcomponentsCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.HasSubcomponentsCard, { variant: "gridItem" }))
  )
});
const catalogHasSystemsEntityCard = createEntityCardExtension({
  name: "has-systems",
  loader: async () => import('../components/HasSystemsCard/index.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.HasSystemsCard, { variant: "gridItem" }))
  )
});
var entityCards = [
  catalogAboutEntityCard,
  catalogLinksEntityCard,
  catalogLabelsEntityCard,
  catalogDependsOnComponentsEntityCard,
  catalogDependsOnResourcesEntityCard,
  catalogHasComponentsEntityCard,
  catalogHasResourcesEntityCard,
  catalogHasSubcomponentsEntityCard,
  catalogHasSystemsEntityCard
];

export { catalogAboutEntityCard, catalogDependsOnComponentsEntityCard, catalogDependsOnResourcesEntityCard, catalogHasComponentsEntityCard, catalogHasResourcesEntityCard, catalogHasSubcomponentsEntityCard, catalogHasSystemsEntityCard, catalogLabelsEntityCard, catalogLinksEntityCard, entityCards as default };
