// @backstage/plugin-techdocs/dist/Router.esm.js
import React from 'react';
import { Routes, Route, useRoutes } from 'react-router-dom';
import { EntityPageDocs } from './EntityPageDocs.esm.js';
import { TechDocsIndexPage } from './home/components/TechDocsIndexPage.esm.js';
import { TechDocsReaderPage } from './reader/components/TechDocsReaderPage/TechDocsReaderPage.esm.js';
import { useEntity, MissingAnnotationEmptyState } from '@backstage/plugin-catalog-react';

const TECHDOCS_ANNOTATION = "backstage.io/techdocs-ref";
const TECHDOCS_EXTERNAL_ANNOTATION = "backstage.io/techdocs-entity";
const isTechDocsAvailable = (entity) => Boolean(entity?.metadata?.annotations?.[TECHDOCS_ANNOTATION]) || Boolean(entity?.metadata?.annotations?.[TECHDOCS_EXTERNAL_ANNOTATION]);
const Router = () => {
  return /* @__PURE__ */ React.createElement(Routes, null, /* @__PURE__ */ React.createElement(Route, { path: "/", element: /* @__PURE__ */ React.createElement(TechDocsIndexPage, null) }), /* @__PURE__ */ React.createElement(
    Route,
    {
      path: "/:namespace/:kind/:name/*",
      element: /* @__PURE__ */ React.createElement(TechDocsReaderPage, null)
    }
  ));
};
const EmbeddedDocsRouter = (props) => {
  const { children } = props;
  const { entity } = useEntity();
  const element = useRoutes([
    {
      path: "/*",
      element: /* @__PURE__ */ React.createElement(EntityPageDocs, { entity }),
      children: [
        {
          path: "*",
          element: children
        }
      ]
    }
  ]);
  const projectId = entity.metadata.annotations?.[TECHDOCS_ANNOTATION] || entity.metadata.annotations?.[TECHDOCS_EXTERNAL_ANNOTATION];
  if (!projectId) {
    return /* @__PURE__ */ React.createElement(MissingAnnotationEmptyState, { annotation: [TECHDOCS_ANNOTATION] });
  }
  return element;
};

export { EmbeddedDocsRouter, Router, isTechDocsAvailable };
