// @backstage/plugin-proxy-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendCommon = require('@backstage/backend-common');
var backendPluginApi = require('@backstage/backend-plugin-api');
var router = require('./cjs/router-BB9HHp8x.cjs.js');
require('express-promise-router');
require('http-proxy-middleware');

var alpha = backendPluginApi.createBackendPlugin({
  pluginId: "proxy",
  register(env) {
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        discovery: backendPluginApi.coreServices.discovery,
        logger: backendPluginApi.coreServices.logger,
        httpRouter: backendPluginApi.coreServices.httpRouter
      },
      async init({ config, discovery, logger, httpRouter }) {
        await router.createRouterInternal({
          config,
          discovery,
          logger: backendCommon.loggerToWinstonLogger(logger),
          httpRouterService: httpRouter
        });
      }
    });
  }
});

exports.default = alpha;
