// @backstage/frontend-plugin-api/dist/extensions/createAppRootWrapperExtension.esm.js
import React from 'react';
import { createExtension } from '../wiring/createExtension.esm.js';
import { createExtensionDataRef } from '../wiring/createExtensionDataRef.esm.js';

function createAppRootWrapperExtension(options) {
  return createExtension({
    kind: "app-root-wrapper",
    namespace: options.namespace,
    name: options.name,
    attachTo: options.attachTo ?? { id: "app/root", input: "wrappers" },
    configSchema: options.configSchema,
    disabled: options.disabled,
    inputs: options.inputs,
    output: {
      component: createAppRootWrapperExtension.componentDataRef
    },
    factory({ inputs, config }) {
      const Component = (props) => {
        return /* @__PURE__ */ React.createElement(options.Component, { inputs, config }, props.children);
      };
      return {
        component: Component
      };
    }
  });
}
((createAppRootWrapperExtension2) => {
  createAppRootWrapperExtension2.componentDataRef = createExtensionDataRef(
    "app.root.wrapper"
  );
})(createAppRootWrapperExtension || (createAppRootWrapperExtension = {}));

export { createAppRootWrapperExtension };
