// @backstage/plugin-search-backend-module-pg/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha$1 = require('@backstage/plugin-search-backend-node/alpha');
var PgSearchEngine = require('./cjs/PgSearchEngine-CWuOiMaJ.cjs.js');
var backendCommon = require('@backstage/backend-common');
require('@backstage/plugin-search-backend-node');
require('uuid');

var alpha = backendPluginApi.createBackendModule({
  pluginId: "search",
  moduleId: "postgres-engine",
  register(env) {
    env.registerInit({
      deps: {
        searchEngineRegistry: alpha$1.searchEngineRegistryExtensionPoint,
        database: backendPluginApi.coreServices.database,
        config: backendPluginApi.coreServices.rootConfig,
        logger: backendPluginApi.coreServices.logger
      },
      async init({ searchEngineRegistry, database, config, logger }) {
        if (await PgSearchEngine.PgSearchEngine.supported(database)) {
          searchEngineRegistry.setSearchEngine(
            await PgSearchEngine.PgSearchEngine.fromConfig(config, {
              database,
              logger: backendCommon.loggerToWinstonLogger(logger)
            })
          );
        } else {
          logger.warn(
            "Postgres search engine is not supported, skipping registration of search-backend-module-pg"
          );
        }
      }
    });
  }
});

exports.default = alpha;
