// @backstage/plugin-scaffolder-node/dist/alpha.cjs.js
'use strict';

var backendPluginApi = require('@backstage/backend-plugin-api');
var tar = require('tar');
var concatStream = require('concat-stream');
var util = require('util');
var stream = require('stream');

function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e : { default: e }; }

var tar__default = /*#__PURE__*/_interopDefaultCompat(tar);
var concatStream__default = /*#__PURE__*/_interopDefaultCompat(concatStream);

const pipeline = util.promisify(stream.pipeline);
const serializeWorkspace = async (opts) => {
  return new Promise(async (resolve) => {
    await pipeline(
      tar__default.default.create({ cwd: opts.path }, [""]),
      concatStream__default.default((buffer) => {
        return resolve({ contents: buffer });
      })
    );
  });
};
const restoreWorkspace = async (opts) => {
  const { buffer, path } = opts;
  if (buffer) {
    await pipeline(
      stream.Readable.from(buffer),
      tar__default.default.extract({
        C: path
      })
    );
  }
};

const scaffolderActionsExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "scaffolder.actions"
});
const scaffolderTaskBrokerExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "scaffolder.taskBroker"
});
const scaffolderTemplatingExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "scaffolder.templating"
});
const scaffolderAutocompleteExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "scaffolder.autocomplete"
});
const scaffolderWorkspaceProviderExtensionPoint = backendPluginApi.createExtensionPoint({
  id: "scaffolder.workspace.provider"
});

exports.restoreWorkspace = restoreWorkspace;
exports.scaffolderActionsExtensionPoint = scaffolderActionsExtensionPoint;
exports.scaffolderAutocompleteExtensionPoint = scaffolderAutocompleteExtensionPoint;
exports.scaffolderTaskBrokerExtensionPoint = scaffolderTaskBrokerExtensionPoint;
exports.scaffolderTemplatingExtensionPoint = scaffolderTemplatingExtensionPoint;
exports.scaffolderWorkspaceProviderExtensionPoint = scaffolderWorkspaceProviderExtensionPoint;
exports.serializeWorkspace = serializeWorkspace;
