// @backstage/plugin-auth-backend-module-guest-provider/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var pluginAuthNode = require('@backstage/plugin-auth-node');
var errors = require('@backstage/errors');
var catalogModel = require('@backstage/catalog-model');

const guestAuthenticator = pluginAuthNode.createProxyAuthenticator({
  defaultProfileTransform: async () => {
    return { profile: {} };
  },
  initialize({ config }) {
    const allowOutsideDev = config.getOptionalBoolean(
      "dangerouslyAllowOutsideDevelopment"
    );
    if (process.env.NODE_ENV !== "development" && allowOutsideDev !== true) {
      throw new errors.NotImplementedError(
        "The guest provider cannot be used outside of a development environment"
      );
    }
  },
  async authenticate() {
    return { result: {} };
  }
});

const signInAsGuestUser = (config) => async (_, ctx) => {
  if (process.env.NODE_ENV !== "development" && config.getOptionalBoolean("dangerouslyAllowOutsideDevelopment") !== true) {
    throw new errors.NotImplementedError(
      "The guest provider is NOT recommended for use outside of a development environment. If you want to enable this, set `auth.providers.guest.dangerouslyAllowOutsideDevelopment: true` in your app config."
    );
  }
  const userRef = config.getOptionalString("userEntityRef") ?? catalogModel.stringifyEntityRef({
    kind: "user",
    namespace: "development",
    name: "guest"
  });
  const ownershipRefs = config.getOptionalStringArray(
    "ownershipEntityRefs"
  ) ?? [userRef];
  try {
    return await ctx.signInWithCatalogUser({ entityRef: userRef });
  } catch (err) {
    return ctx.issueToken({
      claims: {
        sub: userRef,
        ent: ownershipRefs
      }
    });
  }
};

const authModuleGuestProvider = backendPluginApi.createBackendModule({
  pluginId: "auth",
  moduleId: "guest-provider",
  register(reg) {
    reg.registerInit({
      deps: {
        providers: pluginAuthNode.authProvidersExtensionPoint,
        config: backendPluginApi.coreServices.rootConfig
      },
      async init({ providers, config }) {
        providers.registerProvider({
          providerId: "guest",
          factory: pluginAuthNode.createProxyAuthProviderFactory({
            authenticator: guestAuthenticator,
            signInResolver: signInAsGuestUser(
              config.getConfig("auth.providers.guest")
            )
          })
        });
      }
    });
  }
});

exports.default = authModuleGuestProvider;
