// @backstage/plugin-org/dist/alpha.esm.js
import { compatWrapper, convertLegacyRouteRefs } from '@backstage/core-compat-api';
import { createPlugin } from '@backstage/frontend-plugin-api';
import React from 'react';
import { catalogIndexRouteRef } from './routes.esm.js';
import { createEntityCardExtension } from '@backstage/plugin-catalog-react/alpha';

const EntityGroupProfileCard = createEntityCardExtension({
  name: "group-profile",
  filter: "kind:group",
  loader: async () => import('./components/Cards/Group/GroupProfile/GroupProfileCard.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.GroupProfileCard, null))
  )
});
const EntityMembersListCard = createEntityCardExtension({
  name: "members-list",
  filter: "kind:group",
  loader: async () => import('./components/Cards/Group/MembersList/MembersListCard.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.MembersListCard, null))
  )
});
const EntityOwnershipCard = createEntityCardExtension({
  name: "ownership",
  filter: "kind:group,user",
  loader: async () => import('./components/Cards/OwnershipCard/OwnershipCard.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.OwnershipCard, null))
  )
});
const EntityUserProfileCard = createEntityCardExtension({
  name: "user-profile",
  filter: "kind:user",
  loader: async () => import('./components/Cards/User/UserProfileCard/UserProfileCard.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.UserProfileCard, null))
  )
});
var alpha = createPlugin({
  id: "org",
  extensions: [
    EntityGroupProfileCard,
    EntityMembersListCard,
    EntityOwnershipCard,
    EntityUserProfileCard
  ],
  externalRoutes: convertLegacyRouteRefs({
    catalogIndex: catalogIndexRouteRef
  })
});

export { alpha as default };
