// @backstage/plugin-catalog/dist/components/HasComponentsCard/HasComponentsCard.esm.js
import { RELATION_HAS_PART } from '@backstage/catalog-model';
import React from 'react';
import { RelatedEntitiesCard } from '../RelatedEntitiesCard/RelatedEntitiesCard.esm.js';
import { componentEntityColumns, componentEntityHelpLink, asComponentEntities } from '../RelatedEntitiesCard/presets.esm.js';
import { catalogTranslationRef } from '../../translation.esm.js';
import { useTranslationRef } from '@backstage/core-plugin-api/alpha';

function HasComponentsCard(props) {
  const { t } = useTranslationRef(catalogTranslationRef);
  const { variant = "gridItem", title = t("hasComponentsCard.title") } = props;
  return /* @__PURE__ */ React.createElement(
    RelatedEntitiesCard,
    {
      variant,
      title,
      entityKind: "Component",
      relationType: RELATION_HAS_PART,
      columns: componentEntityColumns,
      emptyMessage: t("hasComponentsCard.emptyMessage"),
      emptyHelpLink: componentEntityHelpLink,
      asRenderableEntities: asComponentEntities
    }
  );
}

export { HasComponentsCard };
